### DISCUSSION TITLE
List not updating

### DISCUSSION THREAD

**Discussion by HARDY8118** [msiemens/tinydb#443] [Q&A] [answered]
I have a collection as
```json
[{
    "item_name": "item1",
    "contents": [{ "name" : "ABCD" , "count" : 4 }, { "name" : "BCDEF", "count" : 6 }]
}, {
    "item_name": "item2",
    "contents": [{ "name": "CD", "count": 8 },{ "name" : "DEFG", "count" : 9 }]
}]
```
which I want to change to something like
```json
[{
    "item_name": "item1",
    "contents": [{ "name" : "ABCD" , "count" : 4 }, { "name" : "BCDEF", "count" : 6 }]
}, {
    "item_name": "item2",
    "contents": [{ "name": "CD", "count": 8 },{ "name" : "F", "count" : 9 }]
}]
```
I am trying to change the array using a map function and then updating the list to the result of map operation in the database. 

I am using the following python code

```python
def check(_item):
    _item['name'] = 'F'
    return _item

def _check(_list, _item):
    _query=Query()
    try:
        _list = db.search(_query['item_name'] == _list)[0]
        _list['contents'] = list(map(lambda i: (check(i) if re.match(('^'+_item), i['name']) else i), _list['contents']))
        print(_list['contents'])
        db.update({'contents': _list['contents']}, _query['item_name'] == _list)
        print(db.all())
    except IndexError:
        cprint('List '+args[0]+' does not exists', 'red')

_check('item2', 'D')
```
which produces following output

`[{'name': 'CD', 'count': 8}, {'name': 'F', 'count': 9}]`

`[{'item_name': 'item1', 'contents': [{'name': 'ABCD', 'count': 4}, {'name': 'BCDEF', 'count': 6}]}, {'item_name': 'item2', 'contents': [{'name': 'CD', 'count': 8}, {'name': 'DEFG', 'count': 9}]}]`

The instance of `_list['contents']` seems to be changed successfully after map, but the changes are not reflected in the database.

Please let me know if I am doing anything wrong or some other way to fix this. Any kind of help will be appreciated.

**Comment by msiemens** [comment, marked as answer]
Hey @HARDY8118,

I tested this and it seems like the following problem: You use `_list` to pass the `item_name` you want to modify. But then you overwrite the `_list` variable with the search result. Thus, the second query (`_query['item_name'] == _list`) searches for all documents where `item_name` is equal to `{'item_name': 'item2', 'contents': [{'name': 'CD', 'count': 8}, {'name': 'F', 'count': 9}]}` which I think is _not_ what you wanted 🙂 

This seems to work correctly if modified like this:

```python
def check(_item):
    _item['name'] = 'F'
    return _item

def _check(item_name, _item):
    _query=Query()
    try:
        _list = db.search(_query['item_name'] == item_name)[0]
        _list['contents'] = list(map(lambda i: (check(i) if re.match(('^'+_item), i['name']) else i), _list['contents']))
        print(_list['contents'])
        db.update({'contents': _list['contents']}, _query['item_name'] == item_name)
        print(db.all())
    except IndexError:
        cprint('List '+args[0]+' does not exists', 'red')

_check('item2', 'D')
```

Can you verify this works as expected?

**Reply by HARDY8118** [reply]
Thank you.
That was helpful.
